providers/supermicro/supermicro.go

```diff
@@ -0,0 +1,343 @@
+		Port: "443",
+	}
+
+	for _, opt := range opts {
+		opt(defaultConfig)
+	}
+
+	return &Client{
+		host:   host,
+		user:   user,
+		pass:   pass,
+		port:   defaultConfig.Port,
+		client: httpclient.Build(defaultConfig.httpClientSetupFuncs...),
+		log:    log,
+	}
+}
+
+// Open a connection to a Supermicro BMC using the vendor API.
+func (c *Client) Open(ctx context.Context) (err error) {
+	data := fmt.Sprintf(
+		"name=%s&pwd=%s&check=00",
+		base64.StdEncoding.EncodeToString([]byte(c.user)),
+		base64.StdEncoding.EncodeToString([]byte(c.pass)),
+	)
+
+	headers := map[string]string{"Content-Type": "application/x-www-form-urlencoded"}
+
+	body, status, err := c.query(ctx, "cgi/login.cgi", http.MethodPost, bytes.NewBufferString(data), headers, 0)
+	if err != nil {
+		return errors.Wrap(bmclibErrs.ErrLoginFailed, err.Error())
+	}
+
+	if status != 200 {
+		return errors.Wrap(bmclibErrs.ErrLoginFailed, strconv.Itoa(status))
+	}
+
+	if !bytes.Contains(body, []byte(`url_redirect.cgi?url_name=mainmenu`)) &&
+		!bytes.Contains(body, []byte(`url_redirect.cgi?url_name=topmenu`)) {
+		return errors.Wrap(bmclibErrs.ErrLoginFailed, "unexpected response contents")
+	}
+
+	contentsTopMenu, status, err := c.query(ctx, "cgi/url_redirect.cgi?url_name=topmenu", http.MethodGet, nil, nil, 0)
+	if err != nil {
+		return errors.Wrap(bmclibErrs.ErrLoginFailed, err.Error())
+	}
+
+	if status != 200 {
+		return errors.Wrap(bmclibErrs.ErrLoginFailed, strconv.Itoa(status))
+	}
+
+	token := parseToken(contentsTopMenu)
+	if token == "" {
+		return errors.Wrap(bmclibErrs.ErrLoginFailed, "could not parse CSRF-TOKEN from page")
+	}
+
+	c.csrfToken = token
+
+	return nil
+}
+
+func parseToken(body []byte) string {
+	var key string
+	if bytes.Contains(body, []byte(`CSRF-TOKEN`)) {
+		key = "CSRF-TOKEN"
+	}
+
+	if bytes.Contains(body, []byte(`CSRF_TOKEN`)) {
+		key = "CSRF_TOKEN"
+	}
+
+	if key == "" {
+		return ""
+	}
+
+	re, err := regexp.Compile(`"CSRF_TOKEN", "(?P<token>.*)"`)
+	if err != nil {
+		return ""
+	}
+
+	found := re.FindSubmatch(body)
+	if len(found) == 0 {
+		return ""
+	}
+
+	return string(found[1])
+}
+
+// Close a connection to a Supermicro BMC using the vendor API.
+func (c *Client) Close(ctx context.Context) error {
+	if c.client == nil {
+		return nil
+	}
+
+	_, status, err := c.query(ctx, "cgi/logout.cgi", http.MethodGet, nil, nil, 0)
+	if err != nil {
+		return errors.Wrap(bmclibErrs.ErrLogoutFailed, err.Error())
+	}
+
+	if status != 200 {
+		return errors.Wrap(bmclibErrs.ErrLogoutFailed, strconv.Itoa(status))
+	}
+
+	return nil
+}
+
+// Name returns the client provider name.
+func (c *Client) Name() string {
+	return ProviderName
+}
+
+func (c *Client) Screenshot(ctx context.Context) (image []byte, fileType string, err error) {
+	fileType = "jpg"
+
+	// request screen preview to be saved
+	if err := c.initScreenPreview(ctx); err != nil {
+		return nil, "", err
+	}
+
+	// give the bmc a few seconds to store the screen preview
+	time.Sleep(2 * time.Second)
```

Comment:
ah...this is unfortunate. I imagine this is the best we can do? there's no way programmatically to determine whether the screen preview was stored?

Reply:
Yeah, theres no means to determine if the is ready. The SMC web interface just delays the fetch for a few seconds.